File: Niveau 4/Niveau 4.2/fibre.py
#-------------------------------------------------------------------------------
# Name:        Fibre optique
# Purpose:     France-IOI niveau 4.2
#
# Author:      Sébastien
#
# Created:     28/01/2023
#-------------------------------------------------------------------------------

import time

# Variables globales
resultat = 0

# Compte le nombre de PC dans chaque branche qui part de 'noeud'
# et retient le maximum.
# Retourne le nombre de PC dans la branche dont 'noeud' est la racine.
def compter(arbre, pere, noeud):
    global resultat
    compteur = 1
    tailleBrancheMax = 0
    for voisin in arbre[noeud]:
        if voisin != pere:
            tailleBranche = compter(arbre, noeud, voisin)
            if tailleBrancheMax < tailleBranche:
                tailleBrancheMax = tailleBranche
            compteur += tailleBranche
    if tailleBrancheMax < len(arbre) - compteur:
        tailleBrancheMax = len(arbre) - compteur
    if resultat > tailleBrancheMax:
        resultat = tailleBrancheMax
    return compteur

def main():
    global resultat
    # Crée l'arbre
    nbNoeuds = int(input())
    voisins = [[] for _ in range(nbNoeuds)]
    for _ in range(nbNoeuds-1):
        noeud1, noeud2 = map(int, input().split())
        voisins[noeud1].append(noeud2)
        voisins[noeud2].append(noeud1)

    resultat = nbNoeuds
    _ = compter(voisins, -1, 0)
    print(resultat)

if __name__ == '__main__':
    start = time.time()
    main()
    end = time.time()
    print("Temps d'exécution :", end-start, "s")

"""
10
1 8
3 4
8 3
2 3
2 6
7 6
5 6
5 9
6 0
"""

File: Niveau 4/Niveau 4.2/fibre.cpp
//------------------------------------------------------------------------------
// Name:        Fibre optique
// Purpose:     France-IOI niveau 4.2
//
// Author:      Sébastien
//
// Created:     29/01/2023
//------------------------------------------------------------------------------

#include <iostream>
#include <vector>
using namespace std;

const int MAX_TAILLE = 100*1000;

vector<int> arbre[MAX_TAILLE];
int nbNoeuds = 0;
int resultat = 0;

// Compte le nombre de PC dans chaque branche qui part de 'noeud'
// et retient le maximum.
// Retourne le nombre de PC dans la branche dont 'noeud' est la racine.
int compter(int pere, int noeud)
{
    int compteur = 1;
    int tailleBrancheMax = 0;
    for (int voisin : arbre[noeud])
    {
        if (voisin != pere)
        {
            int tailleBranche = compter(noeud, voisin);
            if (tailleBrancheMax < tailleBranche) tailleBrancheMax = tailleBranche;
            compteur += tailleBranche;
        }
    }
    if (tailleBrancheMax < nbNoeuds - compteur) tailleBrancheMax = nbNoeuds - compteur;
    if (resultat > tailleBrancheMax) resultat = tailleBrancheMax;
    return compteur;
}

int main()
{
    // Crée l'arbre.
    cin >> nbNoeuds;
    for (int i = 0; i < nbNoeuds-1; i++)
    {
        int noeud1, noeud2;
        cin >> noeud1 >> noeud2;
        arbre[noeud1].push_back(noeud2);
        arbre[noeud2].push_back(noeud1);
    }
    
    resultat = nbNoeuds;
    compter(-1, 0);
    cout << resultat << endl;
}
